external-crates/move/crates/move-regex-borrow-graph/src/regex.rs

```diff
@@ -0,0 +1,343 @@
+                regex: self,
+                idx: 0,
+            }
+        }
+    }
+}
+
+impl<Lbl> Extension<Lbl> {
+    /// If self = pq, then remove_prefix(p) returns Some(q) for all possible q
+    pub fn remove_prefix(&self, p: &Regex<Lbl>) -> Vec<Regex<Lbl>>
+    where
+        Lbl: Clone,
+        Lbl: Eq,
+    {
+        let mut p_walk = p.walk();
+        match p_walk.peek() {
+            WalkPeek::EmptySet => {
+                debug_assert!(false);
+                vec![]
+            }
+            WalkPeek::Epsilon => {
+                // p = epsilon so q = self
+                vec![self.clone().into_regex()]
+            }
+            WalkPeek::Label(l_p) => match self {
+                Extension::Epsilon => {
+                    // cannot remove l1 from epsilon
+                    vec![]
+                }
+                Extension::Label(l_self) => {
+                    if l_p != l_self {
+                        // cannot remove l_p if it doesn't match l_self
+                        vec![]
+                    } else {
+                        p_walk.next();
+                        match p_walk.peek() {
+                            WalkPeek::EmptySet => {
+                                debug_assert!(false);
+                                vec![]
+                            }
+                            WalkPeek::Epsilon | WalkPeek::DotStar => {
+                                // p = l_p and q = epsilon (or the case where dot star is epsilon)
+                                vec![Regex::epsilon()]
+                            }
+                            WalkPeek::Label(_) => {
+                                // there is another label that we cannot remove from self
+                                vec![]
+                            }
+                        }
+                    }
+                }
+                // we can remove any prefix and still have dot star
+                Extension::DotStar => vec![Regex::dot_star()],
+            },
+            WalkPeek::DotStar => {
+                match self {
+                    Extension::DotStar => vec![Regex::dot_star()],
+                    Extension::Epsilon => {
+                        // Consider the case where p=.*=epsilon so p and q are epsilon
+                        vec![Regex::epsilon()]
+                    }
+                    Extension::Label(_) => {
+                        // Two possibilities
+                        // p = epsilon and q = self
+                        // p = self and q = epsilon
+                        vec![self.clone().into_regex(), Regex::epsilon()]
+                    }
+                }
+            }
+        }
+    }
+
+    fn into_regex(self) -> Regex<Lbl> {
+        match self {
+            Extension::Epsilon => Regex::epsilon(),
+            Extension::Label(lbl) => Regex::label(lbl),
+            Extension::DotStar => Regex::dot_star(),
+        }
+    }
+}
+
+enum Walk<'a, Lbl> {
```

Comment:
Consider moving all of this inside of `remove_prefix`

Reply:
I can't because I have two different (but similar) implementations